### PR TITLE
Langchain integration

### PR DESCRIPTION
**Pull request by pseudotensor**
- [x] basic isolated demo with OpenAI or HF/sentence_transformers
- [x] use better embedding and LLM (ours) to see if comparable to OpenAI
- [x] incorporate into chatbot with PDF upload
- [x] If well-chunked, extra chunking is not good idea, except to keep below (perhaps) slightly larger limit than chunk_size
- [x] Add pre-dB building for in-memory for now, not when user asks for response
- [x] Avoid re-downloading github etc. each time
- [x] Make chroma work, need to understand how to avoid dup docs and why fails when just load (missing items)
- [x] How to choose cut_distanct?
- [x] Context not added from chat when find sources.  Maybe that's ok-ish, but need to do better.
- [x] Add docs: https://huggingface.co/databricks/dolly-v2-12b#langchain-usage
- [x] https://huggingface.co/datasets/Cohere/wikipedia-22-12/viewer/en/train
- [x] Fix context so doesn't do hard chop, do at human-bot instead.

Future PRs:

- [ ] Make URL getting work
- [ ] Can ground report more by showing matching words as well as matching n-grams from actual sources
- [ ] Add ability to do new separate chat and go back to other chats
- [ ] Add ability to download chat(s).
- [ ] Consume url from chat itself and do look up
- [ ] https://github.com/h2oai/makersaurus
- [ ] Make All just mean look at all dbs
- [ ] Allow any github link to ingest if not on HF.      If on HF, default to h2oGPT and DAI docs.
- [ ] Unsure how to work around https://github.com/chroma-core/chroma/issues/412#issuecomment-1547027011
- [ ] Good or Bad for Flag, and Compare should be which is better

### CODE REVIEW

**Comment by pseudotensor** [comment]
Using HF and our 6.9B model, gives consistent answers.
```
query: Which config.toml enables pytorch for NLP?
answer: The following config.toml file enables PyTorch for NLP:

```
[nlp]
enable_pytorch_nlp_model = true
enable_pytorch_nlp_transformer = true
```

This file can be found in the `config` directory of the DriverlessAI documentation.
sources: ['nlp_settings.rst', 'nlp.rst', 'nlp_settings.rst', 'nlp_settings.rst']
```

Same direct prompt in llama does well too:

image

**Comment by pseudotensor** [comment]
image

**Comment by pseudotensor** [comment]
```

NO caching: Disabled, too greedy even if change model etc.
The model 'OptimizedModule' is not supported for . Supported models are ['BartForCausalLM', 'BertLMHeadModel', 'BertGenerationDecoder', 'BigBirdForCausalLM', 'BigBirdPegasusForCausalLM', 'BioGptForCausalLM', 'BlenderbotForCausalLM', 'BlenderbotSmallForCausalLM', 'BloomForCausalLM', 'CamembertForCausalLM', 'CodeGenForCausalLM', 'CpmAntForCausalLM', 'CTRLLMHeadModel', 'Data2VecTextForCausalLM', 'ElectraForCausalLM', 'ErnieForCausalLM', 'GitForCausalLM', 'GPT2LMHeadModel', 'GPT2LMHeadModel', 'GPTBigCodeForCausalLM', 'GPTNeoForCausalLM', 'GPTNeoXForCausalLM', 'GPTNeoXJapaneseForCausalLM', 'GPTJForCausalLM', 'LlamaForCausalLM', 'MarianForCausalLM', 'MBartForCausalLM', 'MegaForCausalLM', 'MegatronBertForCausalLM', 'MvpForCausalLM', 'OpenAIGPTLMHeadModel', 'OPTForCausalLM', 'PegasusForCausalLM', 'PLBartForCausalLM', 'ProphetNetForCausalLM', 'QDQBertLMHeadModel', 'ReformerModelWithLMHead', 'RemBertForCausalLM', 'RobertaForCausalLM', 'RobertaPreLayerNormForCausalLM', 'RoCBertForCausalLM', 'RoFormerForCausalLM', 'Speech2Text2ForCausalLM', 'TransfoXLLMHeadModel', 'TrOCRForCausalLM', 'XGLMForCausalLM', 'XLMWithLMHeadModel', 'XLMProphetNetForCausalLM', 'XLMRobertaForCausalLM', 'XLMRobertaXLForCausalLM', 'XLNetLMHeadModel', 'XmodForCausalLM'].
```

**Comment by pseudotensor** [comment]
working links:

image

**Comment by pseudotensor** [comment]
streaming working now:

https://github.com/h2oai/h2ogpt/assets/2249614/0c864278-724f-4d90-9f66-4ce5dc59a9b2

**Comment by pseudotensor** [comment]
Example where grounding further by highlighting rare words that appear in response but not in sources:

image

**Comment by pseudotensor** [comment]
WIki:

https://python.langchain.com/en/latest/modules/agents/tools/examples/wikipedia.html

https://en.wikipedia.org/wiki/Wikipedia:Database_download
https://meta.wikimedia.org/wiki/Data_dump_torrents
https://alchemy.pub/wikipedia

https://luizvbo.github.io/post/loading-wikipedia-data-with-python/
http://ftp.acc.umu.se/mirror/wikimedia.org/dumps/enwiki/20230401/enwiki-YYYYMMDD-pages-articles-multistream-index.txt.bz2

https://en.wikipedia.org/wiki/Wikipedia:Database_download#English-language_Wikipedia


https://stackoverflow.com/questions/60101844/how-do-i-use-python-to-parse-large-wikipedia-dump-in-xml-bz2-format-using-multi

```
jon@gpu:/data/jon/enwiki-20230401-pages-articles-multistream$ ls -alrth
total 21G
-rw-r--r--  1 jon jon  21G May 11 19:50 enwiki-20230401-pages-articles-multistream.xml.bz2
drwx------ 85 jon jon 4.0K May 11 22:06 ../
drwx------  2 jon jon 4.0K May 11 22:26 ./
-rw-------  1 jon jon 239M May 11 22:26 enwiki-20230401-pages-articles-multistream-index.txt.bz2
```

https://luizvbo.github.io/post/loading-wikipedia-data-with-python/
```
sudo apt install lbzip2
lbzip2 -d enwiki-20230401-pages-articles-multistream-index.txt.bz2

```

**Comment by pseudotensor** [comment]
```
def extract_namespace(page_name, namespace_map):
    title_parts = page_name.split(":", 1)
    if len(title_parts) == 1:
        return 0, normalize_title(page_name)
    else:
        ns_name, split_title = title_parts
        if namespace_map is None:
            print("None for %s" % split_title, flush=True)
            return 0, normalize_title(page_name)
        elif ns_name in namespace_map:
            return namespace_map[ns_name].id, normalize_title(split_title)
        else:
            return 0, normalize_title(page_name)
```
to deal with None:

image

**Comment by pseudotensor** [comment]
Gets confused even when switching to OnlyLLM:

image

image

**Comment by pseudotensor** [comment]
Better now with no sources and control if use data sources for actual question.  Less magical, but can map that intent with another LLM call later.

image

**Comment by pseudotensor** [comment]
Because context has more complete info, json extraction is more useful:

image


compared to llama 30b with only its own output as context:

image

**Comment by pseudotensor** [comment]
https://github.com/mediawiki-utilities/python-mwviews

**Comment by pseudotensor** [comment]
```
==================================================================================== 4 passed, 1 skipped in 248.67s (0:04:08) =====================================================================================
(h2ollm) jon@pseudotensor:~/h2o-llm$ 

```

**Comment by pseudotensor** [comment]
Direct item in query can be low match just because one of the words is more common.  Should be smaller distance if higher-ngram matches:

image

**Comment by pseudotensor** [comment]
image

image

File: create_data.py [hunk 1]
```diff
@@ -124,21 +133,33 @@ def get_sentences(blob, length):
...
+        import zipfile
+        with zipfile.ZipFile(path_to_zip_file, 'r') as zip_ref:
+            zip_ref.extractall(path)
+        path = os.path.join(path, 'docs/**/*')
+
+    if path is None:
+        if os.path.isdir(os.path.join(home, 'h2oai')):
+            path = os.path.join(home, "h2oai/docs/**/*")
+        else:
+            assert os.path.isdir(os.path.join(home, 'h2oai.superclean')), '%s does not exist' % path
```
[the remark this answers is not in the document]
Comment (pseudotensor) [reply create_data.py:163]: The default for function get_dai_docs()  is now get_pickle=True, and downloads directly from HF, so no need for local source normally unless we want to update what is on HF.  I think that's ok approach.

In addition, the db file can also be obtained directly, without any need to ever reach get_dai_docs.

And generate by default does not enable DAI docs, so I think ok on many levels.

**Review by arnocandel — APPROVED** [review]

**Comment by pseudotensor** [comment]
```
============================================================================================= short test summary info =============================================================================================
FAILED tests/test_manual_test.py::test_chat_context - NotImplementedError: MANUAL TEST FOR NOW
=================================================================== 1 failed, 13 passed, 7 skipped, 1 xpassed, 9 warnings in 364.21s (0:06:04) ====================================================================
(h2ollm) jon@pseudotensor:~/h2o-llm$ 
```

and tested manual test manually.

**Pull request merged**